cmd/zpool/zpool_main.c

```diff
@@ -6909,6 +6920,15 @@ zpool_do_online(int argc, char **argv)
 					(void) printf(gettext("use 'zpool "
 					    "replace' to replace devices "
 					    "that are no longer present\n"));
+				if ((flags & ZFS_ONLINE_EXPAND) &&
+				    oldstate >= VDEV_STATE_DEGRADED) {
```

Comment:
At the risk of being pedantic, I have to point out that my understanding of this `oldstate` check is that you don't want to exit with an error code if the command "partially" succeeded (by successfully getting the drive online). While I agree that we should not exit with an error code here if the `-e` flag was not provided, the fact that a device expansion *was requested* and was not successful seems to merit some form of user notification.

Reply:
So if I understand correctly, your point is that if the expand flag was passed and we're not in HEALTHY, we should always print the warning, but preserve the existing distinction for whether an error code is returned or not (which, to be clear, is "if the device was previously offline, we consider this a successful online even if it isn't a successful expand, but if it was previously online in some form we consider this a failed expand"). I'm fine with those semantics.